src/server/actions/coder.actions.ts

```diff
@@ -0,0 +1,230 @@
+            // GET /users/{user}/workspace/{workspacename}
+            const response = await fetch(`${CODER_API_ENDPOINT}/api/v2/users/${username}/workspace/${workspaceName}`, {
+                method: 'GET',
+                headers: {
+                    Accept: 'application/json',
+                    'Coder-Session-Token': CODER_TOKEN,
+                },
+            })
+
+            // Return true if workspace exists (200 OK), false if not found (404)
+            return {
+                exists: response.ok,
+                data: await response.json(),
+            }
+        } catch (error) {
+            console.error('Error checking workspace existence:', error)
+            // If there's an error (other than 404), we'll consider it as workspace not existing
+            return { exists: false, data: undefined }
+        }
+    })
+
+export const createUserAndWorkspaceAction = new Action('createUserAndWorkspaceAction', { performsMutations: true })
+    .params(
+        z.object({
+            name: z.string().nonempty(),
+            userId: z.string().nonempty(),
+            email: z
+                .string()
+                .nonempty()
+                .regex(/^[^\s@]+@[^\s@]+\.[^\s@]+$/, 'Invalid email address'),
+            studyId: z.string().nonempty(),
+        }),
+    )
+    .handler(async ({ params: { name, userId, email, studyId } }) => {
+        // Load environment variables
+        const CODER_API_ENDPOINT = process.env.CODER_API_ENDPOINT
+        const CODER_TOKEN = process.env.CODER_TOKEN
+        const CODER_TEMPLATE_ID = process.env.CODER_TEMPLATE_ID
+        const CODER_ORGANIZATION = process.env.CODER_ORGANIZATION
+        const username = generateUsername(email, userId)
+
+        if (!CODER_API_ENDPOINT) {
+            throw new Error('CODER_API_ENDPOINT environment variable is not set')
+        }
+        if (!CODER_TOKEN) {
+            throw new Error('CODER_TOKEN environment variable is not set')
+        }
+        if (!CODER_TEMPLATE_ID) {
+            throw new Error('CODER_TEMPLATE_ID environment variable is not set')
+        }
+        if (!CODER_ORGANIZATION) {
+            throw new Error('CODER_ORGANIZATION environment variable is not set')
+        }
+
+        let userData
+
+        try {
+            // First, try to get the user by name
+            const userResponse = await fetch(`${CODER_API_ENDPOINT}/api/v2/users/${username}`, {
+                method: 'GET',
+                headers: {
+                    Accept: 'application/json',
+                    'Coder-Session-Token': CODER_TOKEN,
+                },
+            })
+
+            if (userResponse.ok) {
+                // User exists, get the user data
+                userData = await userResponse.json()
+            } else if (userResponse.status === 400) {
+                // User doesn't exist, create a new one
+                const createUserResponse = await fetch(`${CODER_API_ENDPOINT}/api/v2/users`, {
```

Comment:
Just thinking out loud i wonder if we could do this at a signup/signin step, so that way its just "there", but for this iteration i think its fine here